tree/dataframe/inc/ROOT/RDF/RMergeableValue.hxx

```diff
@@ -0,0 +1,128 @@
+#ifndef ROOT_RDF_RMERGEABLEVALUE
+#define ROOT_RDF_RMERGEABLEVALUE
+
+#include <functional>
+#include <memory>
+
+#include "RtypesCore.h"
+#include "TList.h"
+
+namespace ROOT {
+namespace RDF {
+
+template <typename T>
+class RMergeableValue {
+protected:
+   T fValue;
+
+public:
+   RMergeableValue(const T &value) : fValue{value} {}
+   virtual ~RMergeableValue() = default;
+   virtual void Merge(const std::unique_ptr<RMergeableValue<T>> &) = 0;
+   virtual const T &GetValue() const { return fValue; }
```

Comment:
does this method need to be virtual?

actually, why do we allow users to fetch `fValue`, what's the usecase? Maybe we can limit the interface to just `Merge`

Reply:
oh I see, you need it because `Merge` returns `void`, so you have to call:

```c++
tomerge1->Merge(tomerge2);
tomerge1->Merge(tomerge3);
const auto mergedvalue = tomerge1->GetValue();
```

This API is very easy to use wrong (forget to pass in one of the results to merge, merge 2 into one and then 3 into 2, forget to call `Merge` before you call `GetValue`...).

Something like this would require less boilerplate and would be much harder to get wrong:

```c++
const auto mergedvalue = MergeResultPtrs({res1, res2, res3, res4});
```

Reply:
Because the user actually wants to have their final merged TH1D in the end, right?

Reply:
Probably not virtual though, yes

Reply:
See my second comment in this thread :)

Reply:
I really don't know how I could miss that! sorry :sweat:

Reply:
Solved with the latest API: `MergeValues(std::move(mergeable1), std::move(mergeable2))`